yellowbrick/features/jointplot.py

```diff
@@ -14,308 +14,386 @@
+
+    {joint, hist}_kws : dict, default: None
+        Additional keyword arguments for the plot components.
 
     kwargs : dict
         Keyword arguments that are passed to the base class and may influence
         the visualization as defined in other Visualizers.
 
+    Attributes
+    ----------
+    corr_ : float
+        The correlation or relationship of the data in the joint plot, specified by the
+        correlation algorithm.
+
     Examples
     --------
 
-    >>> visualizer = JointPlotVisualizer()
-    >>> visualizer.fit(X,y)
-    >>> visualizer.poof()
-
-    Notes
-    -----
-    These parameters can be influenced later on in the visualization
-    process, but can and should be set as early as possible.
+    >>> viz = JointPlot(columns=["temp", "humidity"])
+    >>> viz.fit(X, y)
+    >>> viz.poof()
     """
 
-    def __init__(self, ax=None, feature=None, target=None,
-                 joint_plot='scatter', joint_args=None,
-                 xy_plot='hist', xy_args=None,
-                 ratio=5, space=.2, **kwargs):
-
-        # Check matplotlib version - needs to be version 2.0.0 or greater.
-        mpl_vers_maj = int(mpl.__version__.split(".")[0])
-        if mpl_vers_maj < 2:
-            warnings.warn((
-                "{} requires matplotlib major version 2 or greater. "
-                "Please upgrade."
-            ).format(self.__class__.__name__))
-
-        super(JointPlotVisualizer, self).__init__(ax, **kwargs)
-
-        self.feature = feature
-        self.target = target
-        self.joint_plot = joint_plot
-        self.joint_args = joint_args
-        self.xy_plot = xy_plot
-        self.xy_args = xy_args
-        self.ratio = ratio
-        self.space = space
-
-    def fit(self, X, y, **kwargs):
-        """
-        Sets up the X and y variables for the jointplot
-        and checks to ensure that X and y are of the
-        correct data type
-
-        Fit calls draw
+    # TODO: should we couple more closely with Rank2D?
+    correlation_methods = {
+        'pearson': lambda x, y: pearsonr(x,y)[0],
+        'spearman': lambda x, y: spearmanr(x,y)[0],
+        'covariance': lambda x, y: np.cov(x,y)[0,1],
+        'kendalltau': lambda x, y: kendalltau(x,y)[0],
+    }
+
+    def __init__(self, ax=None, columns=None, correlation='pearson', kind="scatter",
+                 hist=True, alpha=0.65, joint_kws=None, hist_kws=None, **kwargs):
+        # Initialize the visualizer
+        super(JointPlot, self).__init__(ax=ax, **kwargs)
+        self._xhax, self._yhax = None, None
+
+        # Set and validate the columns
+        self.columns = columns
+        if self.columns is not None and not isinstance(self.columns, (int, str)):
+            self.columns = tuple(self.columns)
+            if len(self.columns) > 2:
+                raise YellowbrickValueError((
+                    "'{}' contains too many indices or is invalid for joint plot"
```

Comment:
Given my observations with students using different implementations of joint plots, I'm guessing this is going to be one of the more common errors that comes up for users of this visualizer (e.g. providing the wrong combos of column, X, and y args). I'm wondering if we should make this error slightly more verbose (or even break it into two different errors, one for too many indices and another for otherwise invalid input?), so that people have a little bit more to go on.

Reply:
I've expanded the error message - but that does raise an issue; I have validation occurring at both `__init__` and in `fit` e.g. in the case where someone does this:

```python
oz = JointPlot(columns="foo")
oz.columns = ["foo", "baz", "bear"] 
oz.fit(X)
```

The validation in `fit()` is actually more specific because it only occurs as it goes through the possible columns. I wonder if I should just remove the validation in `__init__` and have it only happen in `fit()` and `draw()`? I wonder how sklearn handles this.